src/ModelContextProtocol/Protocol/Transport/StreamableHttpClientSessionTransport.cs

```diff
@@ -0,0 +1,267 @@
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+using ModelContextProtocol.Protocol.Messages;
+using ModelContextProtocol.Utils;
+using ModelContextProtocol.Utils.Json;
+using System.Net.Http.Headers;
+using System.Net.ServerSentEvents;
+using System.Text;
+using System.Text.Json;
+
+namespace ModelContextProtocol.Protocol.Transport;
+
+/// <summary>
+/// The Streamable HTTP client transport implementation
+/// </summary>
+internal sealed partial class StreamableHttpClientSessionTransport : TransportBase
+{
+    private static MediaTypeWithQualityHeaderValue ApplicationJsonMediaType = new("application/json");
+    private static MediaTypeWithQualityHeaderValue TextEventStreamMediaType = new("text/event-stream");
+
+    private readonly HttpClient _httpClient;
+    private readonly SseClientTransportOptions _options;
+    private readonly CancellationTokenSource _connectionCts;
+    private readonly ILogger _logger;
+
+    private string? _mcpSessionId;
+    private Task? _getReceiveTask;
+
+    public StreamableHttpClientSessionTransport(SseClientTransportOptions transportOptions, HttpClient httpClient, ILoggerFactory? loggerFactory, string endpointName)
+        : base(endpointName, loggerFactory)
+    {
+        Throw.IfNull(transportOptions);
+        Throw.IfNull(httpClient);
+
+        _options = transportOptions;
+        _httpClient = httpClient;
+        _connectionCts = new CancellationTokenSource();
+        _logger = (ILogger?)loggerFactory?.CreateLogger<SseClientTransport>() ?? NullLogger.Instance;
+
+        // We connect with the initialization request with the MCP transport. This means that any errors won't be observed
+        // until the first call to SendMessageAsync. Fortunately, that happens internally in McpClientFactory.ConnectAsync
+        // so we still throw any connection-related Exceptions from there and never expose a pre-connected client to the user.
+        SetConnected(true);
+    }
+
+    /// <inheritdoc/>
+    public override async Task SendMessageAsync(
+        JsonRpcMessage message,
+        CancellationToken cancellationToken = default)
+    {
+        using var sendCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, _connectionCts.Token);
+        cancellationToken = sendCts.Token;
+
+        using var content = new StringContent(
+            JsonSerializer.Serialize(message, McpJsonUtilities.JsonContext.Default.JsonRpcMessage),
+            Encoding.UTF8,
+            "application/json"
+        );
+
+        using var httpRequestMessage = new HttpRequestMessage(HttpMethod.Post, _options.Endpoint)
+        {
+            Content = content,
+            Headers =
+            {
+                Accept = { ApplicationJsonMediaType, TextEventStreamMediaType },
```

Comment:
Should we just be caching the whole Headers collection?

Reply:
It turns out that neither Headers nor Headers.Accept are settable, so I had to rely on the collection initializer calling Add. Caching the collection sounds like the type of premature optimization I would normally go for, so I was surprised I hadn't done it already. 😆